lib/grpnice.gd

```diff
@@ -214,151 +337,121 @@ DeclareOperation(
-            if img = fail or
-              not (img in ImagesSource(nice) and
-                PreImagesRepresentative(nice,img)=obj2) then
-                TryNextMethod();
-            fi;
-            return oper( NiceObject(obj1), img );
-        end );
+  InstallNiceMonomorphismMethod( oper, par, "attribute CollElm",
+    IsCollsElms, [ 1 ],
+    function( obj1, obj2 )
+      local nice, img;
+      nice:= NiceMonomorphism( obj1 );
+      img := ImagesRepresentative( nice, obj2 : actioncanfail:= true );
+      if img = fail or
+         not ( img in ImagesSource( nice ) and
+               PreImagesRepresentative( nice, img ) = obj2 ) then
+        TryNextMethod();
+      fi;
+      return oper( NiceObject( obj1 ), img );
+    end );
 end );
 
+
 #############################################################################
 ##
-#F  AttributeMethodByNiceMonomorphismElmColl( <oper>, <par> )
-##
-##  <ManSection>
-##  <Func Name="AttributeMethodByNiceMonomorphismElmColl" Arg='oper, par'/>
-##
-##  <Description>
-##  </Description>
-##  </ManSection>
-##
-BindGlobal( "AttributeMethodByNiceMonomorphismElmColl", function( oper, par )
+#F  AttributeMethodByNiceMonomorphismCollElmOther( <oper>, <par>[, <meth>] )
+##
+BindGlobal( "AttributeMethodByNiceMonomorphismCollElmOther",
+  function( oper, par, meth... )
+  if Length( meth ) = 0 then
+    meth:= function( obj1, obj2, obj3 )
+      local nice, img;
+      nice:= NiceMonomorphism( obj1 );
+      img := ImagesRepresentative( nice, obj2 );
+      return oper( NiceObject( obj1 ), img, obj3 );
+    end;
+  else
+    meth:= meth[1];
+  fi;
+
+  InstallNiceMonomorphismMethod( oper, par, "attribute CollElmOther",
+    true, [ 1 ], meth );
+end );
 
-    # check the argument length
-    if 2 <> Length(par)  then
-        Error( "need two arguments for ", NameFunction(oper) );
-    fi;
-    par    := ShallowCopy(par);
-    par[2] := par[2] and IsHandledByNiceMonomorphism;
 
-    # install the method
-    InstallOtherMethod( oper,
-        "handled by nice monomorphism: attributeElmColl",
-        IsElmsColls,
-        par,
-        0,
-        function( obj1, obj2 )
-            local   nice,img;
-            nice:=NiceMonomorphism(obj2);
-            img := ImagesRepresentative( nice, obj1 );
-            if img = fail or
-              not (img in ImagesSource(nice) and
-                PreImagesRepresentative(nice,img)=obj1) then
-                TryNextMethod();
-            fi;
-            return oper( img,NiceObject(obj2));
-        end );
+#############################################################################
+##
+#F  AttributeMethodByNiceMonomorphismElmColl( <oper>, <par>[, <meth>] )
+##
+BindGlobal( "AttributeMethodByNiceMonomorphismElmColl",
+  function( oper, par, meth... )
+  if Length( meth ) = 0 then
+    meth:= function( obj1, obj2 )
+      local nice, img;
+      nice:= NiceMonomorphism( obj2 );
+      img := ImagesRepresentative( nice, obj1 : actioncanfail:= true );
```

Comment:
This `actioncanfail:= true` is also new and probably desirable / necessary whether we merge this or not, right?

Reply:
It depends.

The idea was that both `AttributeMethodByNiceMonomorphismElmColl` and `AttributeMethodByNiceMonomorphismCollElm` have a special handling of a `fail` result of `ImagesRepresentative`, but only the latter had the `actioncanfail` option.
I think that the behaviour in both situations should be the same.